### DISCUSSION TITLE
Support for several video containers and audio codecs

### DISCUSSION THREAD

**Discussion by ngoquoctoandev** [Tyrrrz/YoutubeExplode#850] [Q&A]
Thank you very much for the useful tool.

I have some doubts, why don't you support some other containers like mkv. In the downloaded video the audio stream contained in it is not set with opus type, instead it is mp4a.

As for the downloaded audio file, if possible why don't you extract both m4a and opus formats.

**Comment by Tyrrrz** [comment]
The library doesn't "support" any specific containers, it just exposes the streams that YouTube provides, which happen to only be in either of two containers: `webm` and `mp4`. Same story for codecs, except there are a bit more permutations.

> As for the downloaded audio file, if possible why don't you extract both m4a and opus formats.

Nothing is stopping you from doing that.

**Comment by ngoquoctoandev** [comment]
@Tyrrrz  When I use the following code to download the video
await youtube.Videos.DownloadAsync(videoUrl, "video.mp4");

It always chooses the best video stream. However for the audio stream it chooses the aac stream instead of the better quality opus stream. If I use the Yt-dlp tool it achieves this on both the video and audio streams.
I would love to get your help in putting the best audio stream in the downloaded video.

**Reply by Tyrrrz** [reply]
It chooses the stream that matches the output container, so `aac` since it's served in `mp4`/`m4a`, while `opus` is served in `webm`. This avoids transcoding and cuts down the processing time because FFmpeg can literally copy data from the input directly over to the output.

If you want to override that behavior, you can manually select the streams as well: https://github.com/Tyrrrz/YoutubeExplode/tree/master/YoutubeExplode.Converter#manually-selecting-streams

**Comment by ngoquoctoandev** [comment]
@Tyrrrz  Instead of creating a new topic, let me ask this library currently only supports progress reporting with double values. I am writing a Console application that uses Spectre's 3rd party library (https://spectreconsole.net/live/progress) to display download progress such as percentage completed, speed, remaining time... Can you support more detailed information to make the download data more intuitive? Thank you very much.

**Reply by Tyrrrz** [reply]
The double value represents progress in a 0..100% range. If you're downloading a single stream, you can interpolate the number of downloaded bytes by using the stream's content length. If it's any of the other, more complicated operations, then it's not possible.
